NBitcoin/Scripting/PubKeyProvider.cs

```diff
@@ -0,0 +1,311 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+
+namespace NBitcoin.Scripting
+{
+	/// <summary>
+	/// Public key objects in descriptors.
+	/// </summary>
+	public abstract class PubKeyProvider : IEquatable<PubKeyProvider>
+	{
+
+		# region subtypes and constructors
+
+		private PubKeyProvider() {}
+
+		/// <summary>
+		/// Wrapper for other pubkey provider which contains (parent key finger print + relative derivation path to inner Pubkey provider)
+		/// </summary>
+		public class Origin : PubKeyProvider
+		{
+			internal Origin(RootedKeyPath keyOriginInfo, PubKeyProvider inner)
+			{
+				if (keyOriginInfo == null)
+					throw new ArgumentNullException(nameof(keyOriginInfo));
+				if (inner == null)
+					throw new ArgumentNullException(nameof(inner));
+				if (inner is Origin)
+					throw new ArgumentException($"Origin can not have {inner} as inner value");
+				KeyOriginInfo = keyOriginInfo;
+				Inner = inner;
+			}
+
+			public RootedKeyPath KeyOriginInfo { get; }
+			public PubKeyProvider Inner { get; }
+
+		}
+
+
+		public class Const : PubKeyProvider
+		{
+			internal Const(PubKey pk)
+			{
+				if (pk == null)
+					throw new ArgumentNullException(nameof(pk));
+				Pk = pk;
+			}
+
+			public PubKey Pk { get; }
+		}
+
+		public enum DeriveType
+		{
+			NO,
+			UNHARDENED,
+			HARDENED
+		}
+		public class HD : PubKeyProvider
+		{
+			public HD(BitcoinExtPubKey extkey, KeyPath path, DeriveType derive)
+			{
+				if (extkey == null)
+					throw new ArgumentNullException(nameof(extkey));
+
+				if (path == null)
+					throw new ArgumentNullException(nameof(path));
+
+				Extkey = extkey;
+				Path = path;
+				Derive = derive;
```

Comment:
Here it could be convenient to check the `derive` value is a valid `DeriveType`. A wrong value here could result in misbehaviors and derivation of wrong keys. 

```c#
			if (!Enum.IsDefined(typeof(DeriveType), value))
			{
				throw new ArgumentException("Invalid value for DeriveType", nameof()derive);
			}
```